src/battery_statefs.c

```diff
@@ -0,0 +1,136 @@
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+ * Authors:
+ *   rinigus <http://github.com/rinigus>
+
+ Battery stats are collected from StateFS Battery namespace. Reported
+ units are as follows:
+
+ capacity %
+ charge %
+ current A
+ energy Wh
+ power W
+ temperature C
+ timefull and timelow seconds
+ voltage V
+
+ Provider at
+ https://git.merproject.org/mer-core/statefs-providers/blob/master/src/power_udev/provider_power_udev.cpp
+
+ **/
+
+#include "collectd.h"
+#include "common.h"
+#include "plugin.h"
+
+#include <stdio.h>
+
+#define STATEFS_ROOT "/run/state/namespaces/Battery/"
+#define BFSZ 512
+
+static int submitted_this_run = 0;
+
+static void battery_submit(const char *type, gauge_t value) {
+  value_t values[1];
+  value_list_t vl = VALUE_LIST_INIT;
+
+  values[0].gauge = value;
+
+  vl.values = values;
+  vl.values_len = 1;
+  sstrncpy(vl.host, hostname_g, sizeof(vl.host));
+  sstrncpy(vl.plugin, "battery", sizeof(vl.plugin));
+  /* statefs supports 1 battery at present */
+  sstrncpy(vl.plugin_instance, "0", sizeof(vl.plugin_instance));
+  sstrncpy(vl.type, type, sizeof(vl.type));
+
+  plugin_dispatch_values(&vl);
+
+  submitted_this_run++;
+}
+
+static _Bool getvalue(const char *fname, gauge_t *value, char *buffer,
+                      size_t buffer_size) {
+  FILE *fh;
+  if ((fh = fopen(fname, "r")) == NULL)
+    return (0);
+
+  if (fgets(buffer, buffer_size, fh) == NULL) {
+    fclose(fh);
+    return (0); // empty file
+  }
+
+  (*value) = atof(buffer);
+
+  fclose(fh);
+
+  return (1);
+}
+
+/* cannot be static, is referred to from battery.c */
+int battery_read_statefs(void) {
+  char buffer[BFSZ];
+  gauge_t value = NAN;
+
+  submitted_this_run = 0;
+
+  if (getvalue(STATEFS_ROOT "ChargePercentage", &value, buffer, BFSZ))
+    battery_submit("charge", value);
+  // Use capacity as a charge estimate if ChargePercentage is not available
+  else if (getvalue(STATEFS_ROOT "Capacity", &value, buffer, BFSZ))
+    battery_submit("charge", value);
+
+  if (getvalue(STATEFS_ROOT "Current", &value, buffer, BFSZ))
+    battery_submit("current", value * 1e-6); // from uA to A
+
+  if (getvalue(STATEFS_ROOT "Energy", &value, buffer, BFSZ))
+    battery_submit("energy", value * 1e-6); // from uWh to Wh
```

Comment:
Sure. its just that with batteries they use Wh. Let's use energy_wh and keep energy for the values in Joule.